### PR TITLE
Reduce intermediate collections when checking symboluse

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

File: src/fsharp/service/service.fs [hunk 1]
```diff
@@ -1855,22 +1855,23 @@ type FSharpCheckProjectResults(projectFileName:string, tcConfigOption, keepAssem
...
 
-        [| for r in tcSymbolUses do yield! r.GetUsesOfSymbol(symbol.Item) |] 
-        |> Seq.distinctBy (fun (itemOcc,_denv,m) -> itemOcc, m) 
-        |> Seq.filter (fun (itemOcc,_,_) -> itemOcc <> ItemOccurence.RelatedText) 
-        |> Seq.map (fun (itemOcc,denv,m) -> FSharpSymbolUse(tcGlobals, denv, symbol, itemOcc, m)) 
+        tcSymbolUses
+        |> Seq.collect (fun r -> r.GetUsesOfSymbol symbol.Item)
+        |> Seq.distinctBy (fun symbolUse -> symbolUse.ItemOccurence, symbolUse.Range) 
+        |> Seq.filter (fun symbolUse -> symbolUse.ItemOccurence <> ItemOccurence.RelatedText) 
+        |> Seq.map (fun symbolUse -> FSharpSymbolUse(tcGlobals, symbolUse.DisplayEnv, symbol, symbolUse.ItemOccurence, symbolUse.Range))
```
Comment (7sharp9) [inline src/fsharp/service/service.fs]: A `choose` would slot in here, Im not sure how many iterations occurs in the global section there always a few hundred to iterate over.

Comment (forki) [reply src/fsharp/service/service.fs]: yeah wsn't sure if it's worth it here

File: src/fsharp/service/service.fs [hunk 2]
```diff
@@ -1855,22 +1855,23 @@ type FSharpCheckProjectResults(projectFileName:string, tcConfigOption, keepAssem
...
+        tcSymbolUses
+        |> Seq.collect (fun r -> r.GetUsesOfSymbol symbol.Item)
+        |> Seq.distinctBy (fun symbolUse -> symbolUse.ItemOccurence, symbolUse.Range) 
+        |> Seq.filter (fun symbolUse -> symbolUse.ItemOccurence <> ItemOccurence.RelatedText) 
+        |> Seq.map (fun symbolUse -> FSharpSymbolUse(tcGlobals, symbolUse.DisplayEnv, symbol, symbolUse.ItemOccurence, symbolUse.Range)) 
         |> Seq.toArray
         |> async.Return
 
     // Not, this does not have to be a SyncOp, it can be called from any thread
     member info.GetAllUsesOfAllSymbols() =
```
Comment (7sharp9) [inline src/fsharp/service/service.fs]: This was always a really slow function, I often wondered if an additional filter could be used for the consumers to limit the returned symbols.  Its used in semantic highlighting in vs4mac and in ionide too for codelens I think.

Comment (forki) [reply src/fsharp/service/service.fs]: filter to what?

Comment (cartermp) [reply src/fsharp/service/service.fs]: In this issue I noticed that we're getting multiple symbols for the same range:  https://github.com/Microsoft/visualfsharp/issues/4231

Perhaps that's something worth filtering on, but I think it's more complicated than that.

Comment (forki) [reply src/fsharp/service/service.fs]: Well I looked at the call sites and they do distinctby. Maybe we need to check these

**Comment by dsyme** [comment]
This looks good to me.  This indeed is a performance-sensitive function as @7sharp9 mentions

**Comment by dsyme** [comment]
@forki As an aside, the one thing I mentally checked for when reviewing the code was that no additional data was being captured by any of the long-lived objects - e.g. no more data is captured by symbols, or TcNameResolutions or other such things.   e.g. see the line https://github.com/Microsoft/visualfsharp/pull/4442/files#diff-dbb9e59b8ac1e1e04e1b00282a91ec97R1458 and the comment before.  

Anyway, all ok - looks good, thanks

**Pull request merged**